partitionmanager/types.py

```diff
@@ -0,0 +1,489 @@
+
+    This kind of partition always resides at the tail of the partition list,
+    and is defined as containing values up to the reserved keyword MAXVALUE.
+    """
+
+    def __init__(self, name, count):
+        self._name = name
+        self._count = count
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def num_columns(self):
+        return self._count
+
+    def values(self):
+        return ", ".join(["MAXVALUE"] * self._count)
+
+    def __lt__(self, other):
+        """MaxValuePartitions are always greater than every other partition."""
+        if isinstance(other, list):
+            if self._count != len(other):
+                raise UnexpectedPartitionException(
+                    f"Expected {self._count} columns but list has {len(other)}."
+                )
+            return False
+        if is_partition_type(other):
+            if self._count != other.num_columns:
+                raise UnexpectedPartitionException(
+                    f"Expected {self._count} columns but list has {other.num_columns}."
+                )
+            return False
+        return ValueError()
+
+    def __eq__(self, other):
+        if isinstance(other, MaxValuePartition):
+            return self.name == other.name and self._count == other.num_columns
+        return False
+
+
+class InstantPartition(PositionPartition):
+    """Represent a partition at the current moment.
+
+    Used for rate calculations as a stand-in that only exists for the purposes
+    of the rate calculation itself.
+    """
+
+    def __init__(self, now, positions):
+        super().__init__("Instant")
+        self._instant = now
+        self._positions = positions
+
+    def timestamp(self):
+        return self._instant
+
+
+class _PlannedPartition(abc.ABC):
+    """Represents a partition this tool plans to emit.
+
+    The method as_partition will make this a concrete type for later evaluation.
+    """
+
+    def __init__(self):
+        self._num_columns = None
+        self._positions = None
+        self._timestamp = None
+        self._important = False
+
+    def set_timestamp(self, timestamp):
+        """Set the timestamp to be used for the modified partition.
+
+        This effectively changes the partition's name.
+        """
+        self._timestamp = timestamp.replace(hour=0, minute=0)
+        return self
+
+    def set_position(self, pos):
+        """Set the position of this modified partition.
+
+        If this partition changes an existing partition, the positions of both
+        must have identical length.
+        """
+        if not isinstance(pos, list):
+            raise ValueError()
+        if self.num_columns is not None and len(pos) != self.num_columns:
+            raise UnexpectedPartitionException(
+                f"Expected {self.num_columns} columns but list has {len(pos)}."
+            )
+        self._positions = pos
+        return self
+
+    def set_important(self):
```

Comment:
What makes a partition important or unimportant?

Reply:
Added some text:
```python
    def set_important(self):
        """Indicate this is an important partition. Used in the
        _plan_partition_changes as a marker that there's a significant
        change in this partition that should be committed even if the
        overall map isn't changing much. """
        self._important = True
        return self
```

Reply:
Ultimately I kinda hate this part, but there's a separation of concerns struggle here where the interior of the algorithm knows this is important, but it also needs to construct scaffolding that is only important if there are subsequent changes.

A refactor of how that all works still eludes me.